File: astrid/common-src/com/todoroo/andlib/sql/Field.java
package com.todoroo.andlib.sql;

import static com.todoroo.andlib.sql.Constants.AND;
import static com.todoroo.andlib.sql.Constants.BETWEEN;
import static com.todoroo.andlib.sql.Constants.COMMA;
import static com.todoroo.andlib.sql.Constants.LEFT_PARENTHESIS;
import static com.todoroo.andlib.sql.Constants.RIGHT_PARENTHESIS;
import static com.todoroo.andlib.sql.Constants.SPACE;

public class Field extends DBObject<Field> {

    protected Field(String expression) {
        super(expression);
    }

    public static Field field(String expression) {
        return new Field(expression);
    }

    public Criterion eq(Object value) {
        if(value == null)
            return UnaryCriterion.isNull(this);
        return UnaryCriterion.eq(this, value);
    }

    public Criterion neq(Object value) {
        if(value == null)
            return UnaryCriterion.isNotNull(this);
        return UnaryCriterion.neq(this, value);
    }

    public Criterion gt(Object value) {
        return UnaryCriterion.gt(this, value);
    }

    public Criterion lt(final Object value) {
        return UnaryCriterion.lt(this, value);
    }

    public Criterion isNull() {
        return UnaryCriterion.isNull(this);
    }

    public Criterion isNotNull() {
        return UnaryCriterion.isNotNull(this);
    }

    public Criterion between(final Object lower, final Object upper) {
        final Field field = this;
        return new Criterion(null) {

            @Override
            protected void populate(StringBuilder sb) {
                sb.append(field).append(SPACE).append(BETWEEN).append(SPACE).append(lower).append(SPACE).append(AND)
                        .append(SPACE).append(upper);
            }
        };
    }

    public Criterion like(final String value) {
        return UnaryCriterion.like(this, value);
    }

    public <T> Criterion in(final T... value) {
        final Field field = this;
        return new Criterion(Operator.in) {

            @Override
            protected void populate(StringBuilder sb) {
                sb.append(field).append(SPACE).append(Operator.in).append(SPACE).append(LEFT_PARENTHESIS);
                for (T t : value) {
                    sb.append(t.toString()).append(COMMA);
                }
                sb.deleteCharAt(sb.length() - 1).append(RIGHT_PARENTHESIS);
            }
        };
    }

    public Criterion in(final Query query) {
        final Field field = this;
        return new Criterion(Operator.in) {

            @Override
            protected void populate(StringBuilder sb) {
                sb.append(field).append(SPACE).append(Operator.in).append(SPACE).append(LEFT_PARENTHESIS).append(query)
                        .append(RIGHT_PARENTHESIS);
            }
        };
    }
}

File: data/src/main/kotlin/com/todoroo/andlib/sql/Field.kt
package com.todoroo.andlib.sql

open class Field(expression: String) : DBObject(expression) {
    fun eq(value: Any?): Criterion = if (value == null) {
        UnaryCriterion.isNull(this)
    } else {
        UnaryCriterion.eq(this, value)
    }

    override fun `as`(newAlias: String): Field {
        val clone = Field(expression)
        clone.alias = newAlias
        return clone
    }

    fun gt(value: Any?): Criterion = UnaryCriterion.gt(this, value)

    fun lt(value: Any?): Criterion = UnaryCriterion.lt(this, value)

    fun lte(value: Any?): Criterion = UnaryCriterion.lte(this, value)

    fun like(value: String?): Criterion = UnaryCriterion.like(this, value)

    fun `in`(query: Query?): Criterion {
        val field = this
        return object : Criterion(Operator.`in`) {
            override fun populate() = "$field IN ($query)"
        }
    }

    fun `in`(entries: List<*>): Criterion {
        val field = this
        return object : Criterion(Operator.`in`) {
            override fun populate() = "$field IN (${entries.joinToString(",")})"
        }
    }

    fun toStringInSelect(): String {
        val sb = StringBuilder(expression)
        if (hasAlias()) {
            sb.append(" AS $alias")
        } else {
            val pos = expression.indexOf('.')
            if (pos > 0 && !expression.endsWith("*")) {
                sb.append(" AS ${expression.substring(pos + 1)}")
            }
        }
        return sb.toString()
    }

    companion object {
        @JvmStatic fun field(expression: String): Field = Field(expression)
        @JvmStatic val COUNT = Field("COUNT(*)")
    }
}